src/tutorial/hello-world/README.md

```diff
@@ -1,2 +1,436 @@
+    
+    0x00 mstore           // place the result in memory
+    0x20 0x00 return      // return the result
+}
+```
+
+The first modifications we make will be within the ADD_TWO macro. On lines 1 and 2 we will shift the calldata offset by 4 bytes for both numbers, this is due to the 4 byte function selector that will be prepended to the calldata value. 
+
+Our `MAIN` macro has changed drastically. 
+The first 4 lines are concerned with isolating the function selector from the calldata.
+1. `0x00` pushed `[0]` onto the stack
+2. `calldataload` takes `[0]` as input and pushes the first 32 bytes of calldata onto the stack
+3. `0xE0` pushes `[224]` onto the stack. This magic number represents 256 bits - 32 bits (32 bytes - 4 bytes). 
+4. When followed by the shr this will place the function selector onto the stack
+
+The following lines will match the function selector on the stack and then jump to the code location where that code exists. Huff handles generating all jump logic for you.
+
+Under the hood the ADD_TWO() macro bytecode will be inlined of ADD_TWO() in the main macro. 
+
+Now you should be able to use libraries like ethers, or other contracts to call your Huff contract! 
+
+We hope this now gives you a good understanding of the main concepts and all of the boiler plate you need to get started in Huff!
+
+## Hello, world!
+Earlier we mentioned how `"Hello, world!"` is quite an advanced concept for Huff. The reason being that we have to clear understanding of EVM ABI string encoding. 
+
+### Primer: ABI Encoding 
+As strings are dynamic types it is not as simple as returning the UTF-8 values for `"Hello, world!"` (`0x48656c6c6f2c20776f726c6421`). In the abi standard, dynamic types are encoded in 3 parts.
+1. The offset of the dynamic data. (A pointer to the start of the dynamic data (uint256))
+2. The length of the dynamic data. (uint256)
+3. The values of the dynamic data. (dynamic length)
+
+Each part will look as follows when returning the string `"Hello, world!"`:
+```
+[Byte number]   [DATA]    
+0x00            0000000000000000000000000000000000000000000000000000000000000020 // The location of the "Hello, world!" data (dynamic type).
+0x20            000000000000000000000000000000000000000000000000000000000000000d // The length of "Hello, world!" in bytes
+0x40            48656c6c6f2c20776f726c642100000000000000000000000000000000000000 // Value "Hello, world!"
+```
+
+Encoding dynamic values takes alot of work!! In order to return `"Hello, world!"` we must return 96 bytes!.
+
+### Implementation
+The following `MAIN` macro steps through this encoding in a clear way (gas optimization will be left as an exercise to the reader!)
+```
+#define macro MAIN() = takes(0) returns(0) {
+    // Store string "Hello, world1" in memory at 0x40
+    0x48656c6c6f2c20776f726c6421        // ["Hello, world!"]
+    0x2d                                // ["Hello, world!", 0x40]
```

Comment:
hmm.  idk.  i think it would b v confusing to a new person to see 0x2d being pushed to stack and referred to as 0x2d.  

Maybe a sidebar above the beginning of this code example that exaplins that the 1st two numbers need to be left padded but the bytes for hello world needs to be right padded.  But if we just MSTORE a number, it will automatically left pad it for us.  So that's why we need to paste it in before 0x40 so that the padding zeros become part of the prior word and the "hello world" is stored in the high bits.

Reply:
Excellent point, I'll add a few words